File: leetcode/78-subsets.py
# https://leetcode.com/problems/subsets/ (top-interview-questions)

from typing import List


class Solution:
    def subsets(self, nums: List[int]) -> List[List[int]]:
        numsSubsets = list()
        numsSubset = list()

        n = (2 ** len(nums)) - 1
        
        while n >= 0:
            for i in range(len(nums)):
                if (n & (1 << i)) != 0:
                    numsSubset.append(nums[i])
            numsSubsets.append(numsSubset.copy())
            numsSubset.clear()
            n -= 1

        return numsSubsets


    def subsets(self, nums: List[int]) -> List[List[int]]:
        numsSubsets = list()
        numsSubset = list()

        def backtracking(index):
            if index == len(nums):
                numsSubsets.append(numsSubset.copy())
                return
            
            numsSubset.append(nums[index])
            backtracking(index+1)
            numsSubset.pop()
            backtracking(index+1)
        
        backtracking(0)
        return numsSubsets

File: leetcode/78-subsets.js
// https://leetcode.com/problems/subsets/ (top-interview-questions)

/**
 * @param {number[]} nums
 * @return {number[][]}
 */
var subsets = function(nums) {
  const subsetList = [];
  const subset = [];

  const backtrack = (i = 0) => {
    if (i === nums.length) {
      subsetList.push([...subset]);
      return;
    }

    backtrack(i+1);
    subset.push(nums[i]);
    backtrack(i+1);
    subset.pop();
  }

  backtrack();

  return subsetList;
};


/**
 * @param {number[]} nums
 * @return {number[][]}
 */
var subsets = function(nums) {
  const subSets = [];
  const subSetCount = (2 ** nums.length) - 1;

  for (let bin = 0; bin <= subSetCount; bin++) {
    const subSet = [];
    for (let i = 0; i < nums.length; i++) {
      if ((bin & (1 << i)) !== 0) {
        subSet.push(nums[i]);
      }
    }
    subSets.push(subSet);
  }

  return subSets;
};
